File: src/graph/Main_1753_최단경로.java
package graph;

import java.io.*;
import java.util.*;

class Node {
	int next, weight;
	
	Node(int next, int weight) {
		this.next = next;
		this.weight = weight;
	}
}

public class Main_1753_최단경로 {

	public static int INF = 987654321;
	public static int ans, V, E, K;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		StringTokenizer st = new StringTokenizer(br.readLine().trim());
		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(br.readLine().trim());
		
		int[] dist = new int[V + 1];
		Arrays.fill(dist, INF);
		
		@SuppressWarnings("unchecked")
		List<Node>[] lines = new List[V + 1];
		for (int i = 1; i <= V; ++i)
			lines[i] = new ArrayList<Node>();
		
		for (int i = 0; i < E; ++i) {
			st = new StringTokenizer(br.readLine().trim());
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());
			int w = Integer.parseInt(st.nextToken());
			
			lines[u].add(new Node(v, w));
		}
		
		Queue<Node> pq = new PriorityQueue<>((n1, n2) -> n1.weight - n2.weight);
		dist[K] = 0;
		pq.add(new Node(K, 0));
		
		while (!pq.isEmpty()) {
			Node node = pq.poll();
			
			for (Node next : lines[node.next]) {
				if (dist[next.next] > dist[node.next] + next.weight) {
					dist[next.next] = dist[node.next] + next.weight;
					pq.add(new Node(next.next, dist[next.next]));
				}
			}
		}
		
		for (int i = 1; i <= V; ++i)
			bw.write(dist[i] == INF ? "INF\n" : dist[i] + "\n");
		
		bw.flush();
		bw.close();
		br.close();
	}

}

File: src/graph/Main_1753_최단경로.cpp
#include <iostream>
#include <vector>
#include <queue>

#define endl    '\n'
#define INF     987654321
#define MAX     20010

using namespace std;

struct node
{
    int next, weight;

    bool operator<(const node& r) const
    {
        return r.weight < weight;
    }
};

int ans, V, E, K;
int dist[MAX];
bool visit[MAX];
vector<pair<int, int>> vec[MAX];
priority_queue<node> pq;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> V >> E >> K;

    fill(dist + 1, dist + V + 1, INF);
    dist[K] = 0;

    for (int i = 0; i < E; ++i)
    {
        int u, v, w;
        cin >> u >> v >> w;
        vec[u].push_back({ v, w });
    }

    pq.push({ K, 0 });

    while (!pq.empty())
    {
        int next = pq.top().next;
        int weight = pq.top().weight;
        pq.pop();

        if (visit[next])
            continue;

        visit[next] = true;

        for (auto next_node : vec[next])
        {
            if (dist[next] + next_node.second < dist[next_node.first])
            {
                dist[next_node.first] = dist[next] + next_node.second;
                pq.push({ next_node.first, dist[next_node.first] });
            }
        }
    }

    for (int i = 1; i <= V; ++i)
    {
        if (dist[i] == INF)
            cout << "INF" << endl;
        else
            cout << dist[i] << endl;
    }
}
